src/librustc_mir/build/mod.rs

```diff
@@ -652,44 +652,86 @@ fn construct_fn<'a, 'gcx, 'tcx, A>(hir: Cx<'a, 'gcx, 'tcx>,
-                var_path: ty::UpvarPath {hir_id: var_hir_id},
-                closure_expr_id: LocalDefId::from_def_id(closure_expr_id),
-            });
-            let by_ref = match capture {
-                ty::UpvarCapture::ByValue => false,
-                ty::UpvarCapture::ByRef(..) => true
-            };
-            let mut decl = UpvarDecl {
-                debug_name: keywords::Invalid.name(),
-                var_hir_id: ClearCrossCrate::Set(var_hir_id),
-                by_ref,
-                mutability: Mutability::Not,
-            };
-            if let Some(Node::Binding(pat)) = tcx.hir().find(var_id) {
-                if let hir::PatKind::Binding(_, _, ident, _) = pat.node {
-                    decl.debug_name = ident.name;
-
-                    if let Some(&bm) = hir.tables.pat_binding_modes().get(pat.hir_id) {
-                        if bm == ty::BindByValue(hir::MutMutable) {
-                            decl.mutability = Mutability::Mut;
-                        } else {
-                            decl.mutability = Mutability::Not;
-                        }
-                    } else {
-                        tcx.sess.delay_span_bug(pat.span, "missing binding mode");
-                    }
-                }
-            }
-            decl
-        }).collect()
-    });
+    let upvar_decls: Vec<_> =  hir_tables.upvar_list[&fn_id].iter().map(|upvar_id| {
+        // fv.var_id() returns the ast::NodeId given the local DefId of a freevar that is being
+        // referred to, by the the closure identified by the fn_id. Our new datastructure
+        // freevar_list can give us all the freevars referred by the closure. Essentially, it gives
+        // the HirId of the freevars (embedded deep within). In this case, to get the equivalent of
+        // what is returned by the var_id(), what we will need is a way to get the NodeId from a
+        // HirId. Question is how we get there. there are utility methods to get the HirId from a
+        // NodeId but not the other way round. var_id is needed because further down, in this
+        // closure, it is used to find the Node::Binding to update the UpvarDecl, which will be
+        // eventually returned by the closure. So the other way would be, if we can get the
+        // Node::Binding, given an HirId, we can also solve this.
+        // let var_id = fv.var_id();
+        let var_hir_id = upvar_id.var_path.hir_id;
+        let capture = hir_tables.upvar_capture(*upvar_id);
+        let by_ref = match capture {
+            ty::UpvarCapture::ByValue => false,
+            ty::UpvarCapture::ByRef(..) => true
+        };
+        let mut decl = UpvarDecl {
+            debug_name: keywords::Invalid.name(),
+            var_hir_id: ClearCrossCrate::Set(var_hir_id),
+            by_ref,
+            mutability: Mutability::Not,
+        };
+        // if let Some(Node::Binding(pat)) = tcx.hir().find(var_id) {
```

Comment:
@blitzerr the simplest fix is to invoke [`tcx.hir().hir_to_node_id`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/map/struct.Map.html#structfield.hir_to_node_id) =)